base_agent/task.py

```diff
@@ -28,56 +25,154 @@ class Task(object):
-        if self.stop_condition.check():
-            self.finished = True
-            return
-        return
+    def check_remove_and_running_children(stepfn):
+        def modified_step(self):
+            if self.remove_condition.check():
+                self.finished = True
+            if self.finished:
+                TaskNode(self.agent.memory, self.memid).get_update_status(
+                    {"prio": -1, "finished": True}
+                )
+                return
+            query = {
+                "base_table": "Tasks",
+                "base_range": {"minprio": 0.5},
+                "triples": [{"pred_text": "_has_parent_task", "obj": self.memid}],
+            }
+            child_task_mems = self.agent.memory.basic_search(query)
+            if child_task_mems:  # this task has active children, step them
+                return
+            return stepfn(self)
+
+        return modified_step
 
-    def add_child_task(self, t, agent, pass_stop_condition=True):
-        """Add a child task to the task_stack and pass along the id 
-        of the parent task (current task)"""
-        # FIXME, this is ugly and dangerous; some conditions might keep state etc?
-        if pass_stop_condition:
-            t.stop_condition = self.stop_condition
-        agent.memory.task_stack_push(t, parent_memid=self.memid)
+    @check_remove_and_running_children
+    def step(self):
+        """The actual execution of a single step of the task is defined here."""
+        pass
 
+    # FIXME remove all this its dead now...
     def interrupt(self):
         """Interrupt the task and set the flag"""
         self.interrupted = True
 
     def check_finished(self):
-        """Check if the task has mark itself finished
+        """Check if the task has marked itself finished
         
         Returns:
             bool: If the task has finished
         """
         if self.finished:
             return self.finished
 
-    def hurry_up(self):
-        """Speed up the task execution"""
-        self.throttling_tick /= 4
-        if self.throttling_tick < THROTTLING_TICK_LOWER_LIMIT:
-            self.throttling_tick = THROTTLING_TICK_LOWER_LIMIT
-
-    def slow_down(self):
-        """Slow down task execution"""
-        self.throttling_tick *= 4
-        if self.throttling_tick > THROTTLING_TICK_UPPER_LIMIT:
-            self.throttling_tick = THROTTLING_TICK_UPPER_LIMIT
+    def add_child_task(self, t):
+        TaskNode(self.agent.memory, self.memid).add_child_task(t)
 
     def __repr__(self):
         return str(type(self))
+
+    check_remove_and_running_children = staticmethod(check_remove_and_running_children)
+
+
+# FIXME new_tasks_fn --> new_tasks
+# FIXME/TODO: name any nonpicklable attributes in the object
+# should we just have this be the behavior of the step of an unsublcassed Task object?
+class ControlBlock(Task):
+    """Container for task control
+    
```

Comment:
More details here since this is a major change. What is the scope of the ControlBlock?

Reply:
not sure what you mean by "scope"?  the control block implements the basic event control (starting and stopping child tasks when conditions are met).  will beef up docstrings